parsers/UY.py

```diff
@@ -46,118 +62,182 @@ def get_salto_grande(session: Optional[Session]) -> float:
-
-        obj[k] = v
+    The function will first fetch data for all avaliable times,
+    then use make_output to format the data points before returning
+    a 24 hr history of data entries
+    """
+    # handle datetime here, so we only have to do it once
+    # get data from webpage
+    raw_entries = parse_page(session)
 
-    # https://github.com/tmrowco/electricitymap/issues/1325#issuecomment-380453296
-    salto_grande = get_salto_grande(session)
-    obj["Hidráulica"] = obj.get("Hidráulica", 0.0) + salto_grande
+    entries = []
+    for raw_entry in raw_entries:
+        if correct_hydro:
+            # https://github.com/tmrowco/electricitymap/issues/1325#issuecomment-380453296
+            raw_entry = correct_for_salto_grande(raw_entry, session)
+        entries.append(make_output(raw_entry))
 
-    return obj
+    return entries
+
+
+def fetch_consumption(
+    zone_key: str = "UY",
+    session: Optional[Session] = Session(),
+    target_datetime: Optional[datetime] = None,
+    logger: Logger = getLogger(__name__),
+) -> Union[dict, List[dict]]:
+
+    if target_datetime is not None:
+        raise ParserException("This parser is unnable to parse dates in the past")
+
+    # helper func to format data output
+    def make_datum(entry):
+        return {
+            "zoneKey": zone_key,
+            "datetime": entry["time"].datetime,
+            "consumption": entry["demand"],
+            "source": "ute.com.uy",
+        }
+
+    # then delegate all the actual work to get_entry_list
+    return get_entry_list(session, make_datum, correct_hydro=False)
 
 
 def fetch_production(
     zone_key: str = "UY",
-    session: Optional[Session] = None,
+    session: Optional[Session] = Session(),
     target_datetime: Optional[datetime] = None,
     logger: Logger = getLogger(__name__),
-) -> dict:
-    if target_datetime:
-        raise NotImplementedError("This parser is not yet able to parse past dates")
+) -> Union[dict, List[dict]]:
 
-    obj = parse_page(session)
+    if target_datetime is not None:
+        raise ParserException("This parser is unnable to parse dates in the past")
 
-    data = {
-        "zoneKey": zone_key,
-        "datetime": obj["datetime"],
-        "production": dict(
-            [(k, obj[INV_MAP_GENERATION[k]]) for k in INV_MAP_GENERATION.keys()]
-        ),
-        "source": "ute.com.uy",
-    }
+    # make a helper function to create output format
+    def make_datum(entry):
+        return {
+            "zoneKey": zone_key,
+            "datetime": entry["time"].datetime,
+            "production": {key: entry[key] for key in AVALIABLE_KEYS},
+            "source": "ute.com.uy",
+        }
 
-    return data
+    # then delegate all the actual work to get_entry_list
+    return get_entry_list(session, make_datum, correct_hydro=True)
 
 
 def fetch_exchange(
     zone_key1: str = "UY",
     zone_key2: str = "BR-S",
-    session: Optional[Session] = None,
+    session: Optional[Session] = Session(),
```

Comment:
Optional can be removed here as well.

Reply:
Same as above, I'm confused because the optional `session` argument seems to be in line with what's recommended in the format defined in the wiki for `fetch_exchange`